Explore time series pretest

exfalsoquodlibet opened the pull request:
EDA notebook of the pre-intervention time-series data to identify characteristics, like seasonality and trends, that will need to be modelled/ accounted for in the interrupted time series analysis.

exfalsoquodlibet commented:
Also addressed the comment you left in the WIP PR around adding annotated text to the vertical (date) lines: https://github.com/alphagov/govuk_ab_analysis/pull/24#issuecomment-993629185

suganya-s commented:
The `y=0` in the text bits means the y axis has to start from 0. I tried to write something (excuse the terrible R) to move it so the axis doesn't get re-sized

```
plot_timeseries <- function(data, ts_var="", title="", x_title=""){
    #'@param data (data.frame) : dataset  
    #'@param ts_var (character string) : name of the variable containing the time-series data
    #'@param title (character string) : plot title
    #'@param x_title (character string) : x-axis title
    #'@return time-series plot
    
    if(!"date" %in% colnames(data)) stop(paste0("column `date` is missing from dataset"))
    
    sym_ts_var <- dplyr::sym(ts_var)
    
    min_y_val <- min(select(data, !!sym_ts_var))
    
    data %>% 
    ggplot2::ggplot(., aes(date, !!sym_ts_var)) +
    geom_point(size=2) +
    geom_line(size=1) +
    #geom_smooth(method="lm", colour="blue") +
    geom_smooth(method = "loess", formula=y~x, colour="red", se=TRUE) +
    geom_vline(aes(xintercept = as.Date("20211111", "%Y%m%d")), col="blue", linetype=2) +
    geom_vline(aes(xintercept = as.Date("20211025", "%Y%m%d")), col="blue", linetype=2) +
    geom_text(aes(x=as.Date("20211111", "%Y%m%d"), y=min_y_val, label="11 Nov 2021"), size=6, angle=90, vjust=-0.4, hjust=0, color="blue") +
    geom_text(aes(x=as.Date("20211025", "%Y%m%d"), y=min_y_val, label="25 Oct 2021"), size=6, angle=90, vjust=-0.4, hjust=0, color="blue") +
    labs(
        title = title,
        subtitle = "Pre-intervention time series") +
    ylab(x_title) +
    theme_custom
    }
 ```

suganya-s commented:
I don't quite get what functions you're using in `plot_SLT` but it all looks good and runs for me. The only thing I may suggest is run it again with the new denominator and then add to this notebook or the Google doc what seasonality we are obviously seeing.

suganya-s reviewed (commented):
comments as above

suganya-s commented on notebooks/2021_tests/gather_metrics_control.sql:
```diff
@@ -0,0 +1,136 @@
+-- create time-series to use as control 
```
I thought I'd widen the net a bit, by counting any visitors that had seen a page with manually curated related links, and then filtering the counting of those that clicked by the content IDs too

```
-- create time-series to use as control 

-- for each day queried, returns counts of:
-- visitors - visitors that viewed at least one page with manually curated related links
-- visitors_that_clicked related links on pages that had manually curated related links
-- visitors_2_or_more related links that day
-- visitors_that_clicked_navigation elements - home link, breadcrumbs, and "Explore this topic" links (on pages with manually curated related links)
-- visitors_that_used_search

-- uses data created by notebooks/2021_tests/get pages with manually curated links.ipynb 
-- that returns all the content items that have manually curated related links, this snapshot was
-- uplaoded to BigQuery as `govuk-bigquery-analytics.datascience.manual_related_links_pages`

CREATE OR REPLACE TABLE `govuk-bigquery-analytics.datascience.control_manual_links_20211011_20211215_data` AS

WITH relevant_fields AS (
    SELECT
    date,
    fullVisitorId,
    visitNumber,
    hits.hitNumber AS hitNumber,
    hits.type AS hit_type,
    hits.page.pagePath AS pagePath,
    hits.eventInfo.eventCategory AS eventCategory,
    hits.eventInfo.eventAction AS eventAction,
    (SELECT
      value
    FROM
      hits.customDimensions
    WHERE
      index=4) AS content_id
    FROM `govuk-bigquery-analytics.87773428.ga_sessions_*` AS sessions
    CROSS JOIN
    UNNEST(sessions.hits) AS hits
    WHERE
     _TABLE_SUFFIX BETWEEN '20211011'
     AND '20211215'
    ),

-- find visitors that viewed a page with manually curated related links
mainstream_rl_page_visitors AS (
    SELECT DISTINCT
        date,
        fullVisitorId
    FROM relevant_fields
    WHERE content_id IN (SELECT content_id FROM `govuk-bigquery-analytics.datascience.manual_related_links_pages`)
),


per_visitor_table AS (
    SELECT
    date,
    fullVisitorId,
    COUNT(DISTINCT 
        CASE WHEN eventCategory = 'relatedLinkClicked'
        AND CONTAINS_SUBSTR(eventAction, 'Related content') 
        AND content_id IN (SELECT content_id FROM `govuk-bigquery-analytics.datascience.manual_related_links_pages`)
        THEN CONCAT(visitNumber,'-',hitNumber) END) AS number_rl_clicked,

    MAX(CASE WHEN (eventCategory IN ('breadcrumbClicked', 'homeLinkClicked')
        OR (eventCategory = 'relatedLinkClicked' 
            AND CONTAINS_SUBSTR(eventAction, 'Explore the topic')))
    AND content_id IN (SELECT content_id FROM `govuk-bigquery-analytics.datascience.manual_related_links_pages`)
        THEN 1 ELSE 0 END) AS navigation_clicked,

    MAX(CASE WHEN hit_type = 'PAGE' AND STARTS_WITH(pagePath, '/search?q=')
        THEN 1 ELSE 0 END) AS search_used,
    FROM
        relevant_fields
    INNER JOIN mainstream_rl_page_visitors USING (date, fullVisitorId)
    GROUP BY
    date,
    fullVisitorId
),

counts AS (
    SELECT 
    date,
    COUNT(DISTINCT fullVisitorId) AS visitors,
    -- shall we also look at proportion of visitors that clicked related links more than once?
    -- indicates the first link was good enough quality for them to not lose confidence in recommendations
    COUNT(DISTINCT CASE WHEN number_rl_clicked > 0 THEN fullVisitorId END) AS visitors_that_clicked_rl,
    COUNT(DISTINCT CASE WHEN number_rl_clicked > 1 THEN fullVisitorId END) AS visitors_2_or_more_rl,
    -- clicking breadcrumbs, a link to the homepage, or "Explore the topic" links 
    -- indicate a user may not be finding what they are looking for and related links aren't helpful to them
    SUM(navigation_clicked) AS visitors_that_clicked_navigation,
    COUNT(DISTINCT CASE WHEN number_rl_clicked > 0 AND navigation_clicked = 0 THEN fullVisitorId END) AS visitors_that_clicked_rl_and_no_nav,
    -- an increase in users searching may indicate that more users are lost and related links aren't helping
    SUM(search_used) AS visitors_that_used_search
    FROM per_visitor_table
    GROUP BY 
    date
)

SELECT
date,
visitors,
visitors_that_clicked_rl,
visitors_2_or_more_rl,
visitors_that_clicked_navigation,
visitors_that_clicked_rl_and_no_nav,
visitors_that_used_search,
-- proportions
ROUND(visitors_that_clicked_rl / visitors, 5) AS pc_visitors_used_rl,
ROUND(visitors_2_or_more_rl / visitors, 5)  AS pc_visitors_2_or_more_rl,
ROUND(visitors_2_or_more_rl / visitors_that_clicked_rl, 5) AS pc_visitors_returning_to_rl,
ROUND(visitors_that_clicked_navigation / visitors, 5)  AS pc_visitors_that_clicked_navigation,
ROUND(visitors_that_used_search / visitors, 5)  AS pc_visitors_that_used_search
FROM
counts
```

exfalsoquodlibet replied:
Thanks Suganya. I'm not convinced (blunt, I know :D).

For the control time-series to work as a control, it must be:
(1) unaffected by the intervention
(2) share potential confounder with the intervention time-series

The problem with this "looser" approach is that we can no longer ensure (1), in that these visitors may have visited pages other than "mainstream manually-curated-rl" pages and even if we are not looking at their behaviour on those pages, the exposure may have still influenced what they do on the "mainstream .." pages.
The stricter definition avoids that and so is a "safer" control.

Happy to discuss offline (in case I have missed something).

suganya-s replied:
That's a good point, thanks for being stringent about this so we can get useful conclusions!

exfalsoquodlibet commented:
> I don't quite get what functions you're using in `plot_SLT` but it all looks good and runs for me. The only thing I may suggest is run it again with the new denominator and then add to this notebook or the Google doc what seasonality we are obviously seeing.

I am decomposing the time-series into the components `trend`, `seasonality` and `random noise` via the `feasts::SLT` function that implements “Seasonal and Trend decomposition using Loess,” here:
```
decomp <- data %>% model(STL(!!sym_ts_var)) %>% components()
```
More info here: https://otexts.com/fpp2/stl.html

This is run with the new denominator now. And I have also added the Conclusions session where it captures the type of day-to-day seasonality we observe. Also added to the doc.

suganya-s commented on notebooks/2021_tests/gather_metrics_control.sql:
```diff
@@ -0,0 +1,136 @@
+            AND CONTAINS_SUBSTR(eventAction, 'Explore the topic')))
+        THEN 1 ELSE 0 END) AS navigation_clicked,
+
+    MAX(CASE WHEN hit_type = 'PAGE' AND STARTS_WITH(pagePath, '/search?q=')
```
I think we need to add this to the "allowed pages" if we want to see this be non-zero

exfalsoquodlibet replied:
Good point. To minimise the risk to the "pureness" of our control time-series, I have now removed the internal-search usage; in any case, it was not used as main metric.
Done in   https://github.com/alphagov/govuk_ab_analysis/pull/25/commits/6651fda95f8b9493c8ccb70201f45c5dd34bb269

suganya-s commented on notebooks/2021_tests/gather_metrics_control.sql:
```diff
@@ -0,0 +1,136 @@
+    SELECT 
+        *,
+        CASE WHEN content_id IN (SELECT content_id FROM `govuk-bigquery-analytics.datascience.manual_related_links_pages`) 
+            OR doc_type IN UNNEST(NO_N2V_DOCS) THEN 1
```
As we talked about yesterday, you can't click on related links from the document types in `NO_N2V_DOCS` so do we want to include them?

exfalsoquodlibet replied:
Yes, god shout. Forgot to push the changes after our meeting yesterday. Done now in https://github.com/alphagov/govuk_ab_analysis/pull/25/commits/6651fda95f8b9493c8ccb70201f45c5dd34bb269

exfalsoquodlibet commented:
Let me know if you are happy for me to merge this to `main` @suganya-s

suganya-s commented on notebooks/2021_tests/gather_metrics_control.sql:
```diff
@@ -0,0 +1,124 @@
+      hits.customDimensions
+    WHERE
+      index=4) AS content_id,
+    (SELECT
```
you don't have to select `doc_type` in this query anymore, save a couple of pennies :)

exfalsoquodlibet replied:
All good points, Suganya. Addressed in https://github.com/alphagov/govuk_ab_analysis/pull/25/commits/eb8aba77828743a33ce7c092481a09b7a3d11e00

suganya-s commented on notebooks/2021_tests/gather_metrics_control.sql:
```diff
@@ -0,0 +1,124 @@
+    visitNumber,
+    hits.hitNumber AS hitNumber,
+    hits.type AS hit_type,
+    hits.page.pagePath AS pagePath,
```
don't think we need pagePath any more either :)

exfalsoquodlibet replied:
Again, thanks for flagging this. Addressed in https://github.com/alphagov/govuk_ab_analysis/pull/25/commits/eb8aba77828743a33ce7c092481a09b7a3d11e00

suganya-s commented on notebooks/2021_tests/gather_metrics_control.sql:
```diff
@@ -0,0 +1,124 @@
+    GROUP BY fullVisitorId, date
+),
+
+purely_non_n2v_visitors AS (
```
I don't think you need `DISTINCT` here as we group by `fullVisitorId, date` in the previous query

exfalsoquodlibet replied:
Done in https://github.com/alphagov/govuk_ab_analysis/pull/25/commits/eb8aba77828743a33ce7c092481a09b7a3d11e00

suganya-s reviewed (approved):
just a few optional suggestions in the SQL, and the moving where the dates get printed  in the plots, but otherwise all good!
